Automatic categories

jenno-verdonck opened the pull request:
I noticed the addition of a ```df_to_pandas``` method with the ```categories``` dictionary as an argument. I added some methods to automatically generate this dictionary using the definitions available [here](https://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/). This functionality is only available for ```year```>=2017 as definition .csv files aren't available for previous years. The code does also not translate ```PUMA``` codes as these are only unique in combination with the state code and require an additional [file](https://www2.census.gov/geo/docs/reference/puma/2010_PUMA_Names.txt).

AndresAlgaba commented on examples/encoding_categorical_features.ipynb:
```diff
@@ -136,6 +136,49 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from folktables import ACSIncome_categories\n",
```
This import does not exist in the latest version of PR #21 which has been merged into the main branch so it can be removed.

AndresAlgaba commented on examples/encoding_categorical_features.ipynb:
```diff
@@ -136,6 +136,49 @@
+    "\n",
+    "ca_features, ca_labels, _ = ACSIncome.df_to_pandas(ca_data, categories=categories, dummies=True)\n",
+    "\n",
+    "ca_features.to_csv('ca_features.csv', index=False)\n",
```
Maybe we can just use .head() instead of .to_csv() as the README already shows how to save the data via pandas. This can also be changed a few cells above in the initial example.

AndresAlgaba commented on examples/encoding_categorical_features.ipynb:
```diff
@@ -136,6 +136,49 @@
+   "source": [
+    "from folktables import ACSIncome_categories\n",
+    "\n",
+    "data_source = ACSDataSource(survey_year='2018', horizon='1-Year', survey='person')\n",
```
data_source and ca_data are already created a few cells above, I do not think this should be repeated.

AndresAlgaba commented on folktables/folktables.py:
```diff
@@ -141,7 +141,45 @@ def df_to_pandas(self, df, categories=None, dummies=False):
+            mapping_dict = {key: value.replace(';', ',') for (key, value) in
+                            zip(mapped_col.tolist(), coll_definition[6].tolist())}
+
+            for key, value in mapping_dict.items():
```
I believe these prints can be removed?

AndresAlgaba commented on folktables/load_acs.py:
```diff
@@ -70,8 +70,7 @@ def initialize_and_download(datadir, state, year, horizon, survey, download=Fals
     remote_fname = f'csv_{survey_code}{state.lower()}.zip'
-    url = os.path.join(base_url, remote_fname).replace('\\', '/')  # Transform Windows path into valid url
-
+    url = f'{base_url}/{remote_fname}'
```
This also works fine for me on Windows :)

AndresAlgaba reviewed (changes requested):
Thank you for adding this convenient method! It was exactly what I was looking for as a follow-up! I proposed some changes via comments, and was wondering whether you could allow users to maybe specify which features they want to encode by adding a ```features``` argument to the ```generate_categories``` method? The default could be ```None``` in which case all the features are encoded as is the case now.

jenno-verdonck commented:
@AndresAlgaba Thank you for the suggestions. I applied all the changes you suggested to the code.

AndresAlgaba reviewed (approved):
Glad the suggestions were helpful! Thanks again for the effort :) LGTM!

mrtzh commented on folktables/folktables.py:
```diff
@@ -141,7 +141,42 @@ def df_to_pandas(self, df, categories=None, dummies=False):
         return variables, target, group
 
-    @property 
+    def generate_categories(self, definition_df):
```
If I understand correctly, this function is specific to ACS PUMS. In that case, it should not go into `folktables.py` and especially not as a method of `BasicProblem`. This file contains general code that is not specific to a single data source. Likewise the class `BasicProblem` needs to be data source agnostic.

I suggest we move this method into `load_acs.py`. Does this make sense?

jenno-verdonck replied:
Thank you for the feedback. I understand that the current location of the `generate_categories` method is not ideal. My reasoning was that this code uses the `features` attribute of the specific problem and is thus problem specific.
I can move this functionality by requiring the user to submit the features themselves.
I would however suggest to move this into `acs.py` instead of `load_acs.py`. `load_acs.py` seems to have no exposed methods to a user and seems to mainly contain code downloading files from the Census website.

Are you fine with this approach?

jenno-verdonck replied:
I have currently placed it in ```load_acs.py``` and exposed it to the user in the ```_init_.py``` file.

mrtzh commented on folktables/folktables.py:
```diff
@@ -141,7 +141,42 @@ def df_to_pandas(self, df, categories=None, dummies=False):
+    def generate_categories(self, definition_df):
+        """
+        Generates a categories dictionary using the provided definition dataframe. Does not create a category mapping
+        for variable requiring the 2010 Public use microdata area code (PUMA) as these need an additional definition
```
Clarify in the doc string what `definition_df` is and what assumptions you make about this dataframe. See other methods for how to describe arguments in a doc string.

mrtzh commented on folktables/folktables.py:
```diff
@@ -141,7 +141,42 @@ def df_to_pandas(self, df, categories=None, dummies=False):
+                # do not add to categories
+                continue
+
+            # transform to numbers as definitions are strings.
```
The following section could use better documentation. I find it hard to understand what's happening here. Is there a way to do it without using the `-99999999999999.0` trick?

mrtzh replied:
In addition, would you be able to add some tests for what's being implemented here? Thanks!

mrtzh reviewed (changes requested):
Hi! I really appreciated the contribution. I'd love to see a few changes, if possible, to make sure that this code meets the requirements of the main package codebase. Please see my comments for suggestions. Thanks!

jenno-verdonck commented:
> Hi! I really appreciated the contribution. I'd love to see a few changes, if possible, to make sure that this code meets the requirements of the main package codebase. Please see my comments for suggestions. Thanks!

Thanks for the review. I pushed some changes addressing your comments.
The ```generate_categories``` method has been relocated to the ```load_acs.py``` file. This results in the user having to define the features included in the categories dictionary. I also had to include the ```generate_categories``` in the ```__init__.py``` file to improve user access to the function.

The ```-99999999999999.0``` trick has been simplified but will still be needed if a placeholder definition for non-numeric values is desired (some attributes define there own definition, others don't). I added some comments to clarify the reason behind this trick.

Finaly I improved the docstring and added some basic tests.
